domain/secret/state/state.go

```diff
@@ -357,30 +415,277 @@ func (st State) createSecret(
+			return errors.Annotatef(err, "deleting next rotate record for secret %q", uri)
+		}
+	}
+
+	if len(secret.Data) == 0 && secret.ValueRef == nil {
+		return nil
+	}
+
+	revisionUUID, err := uuid.NewUUID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	nextRevision := existing[0].LatestRevision + 1
+	dbRevision := secretRevision{
+		ID:         revisionUUID.String(),
+		SecretID:   uri.ID,
+		Revision:   nextRevision,
+		CreateTime: now,
+		UpdateTime: now,
+	}
+	if err := st.upsertSecretRevision(ctx, tx, dbRevision, secret.ExpireTime); err != nil {
+		return errors.Annotatef(err, "inserting revision for secret %q", uri)
+	}
+
+	if len(secret.Data) > 0 {
+		if err := st.updateSecretContent(ctx, tx, dbRevision.ID, secret.Data); err != nil {
+			return errors.Annotatef(err, "updating content for secret %q", uri)
+		}
+	}
+
+	if secret.ValueRef != nil {
+		if err := st.upsertSecretValueRef(ctx, tx, dbRevision.ID, secret.ValueRef); err != nil {
+			return errors.Annotatef(err, "updating backend value reference for secret %q", uri)
+		}
+	}
+
+	if err := st.markObsoleteRevisions(ctx, tx, uri, nextRevision); err != nil {
+		return errors.Annotatef(err, "marking obsolete revisions for secret %q", uri)
+	}
+
+	if dbSecret.AutoPrune {
+		if err := st.pruneUnusedRevisions(ctx, tx, uri); err != nil {
+			return errors.Annotatef(err, "deleting unused revisions for secret %q", uri)
+		}
+	}
+	return nil
+}
+
+func (st State) upsertSecretLabel(ctx context.Context, tx *sqlair.TX, uri *coresecrets.URI, label string, owner coresecrets.Owner) error {
+	switch owner.Kind {
+	case coresecrets.ModelOwner:
+		dbSecretOwner := secretModelOwner{
+			SecretID: uri.ID,
+			Label:    label,
+		}
+		if err := st.upsertSecretModelOwner(ctx, tx, dbSecretOwner); err != nil {
+			return errors.Annotatef(err, "updating model secret record for secret %q", uri)
+		}
+	case coresecrets.ApplicationOwner:
+		dbSecretOwner := secretApplicationOwner{
+			SecretID: uri.ID,
+			// Query selects the application uuid as owner id.
+			ApplicationUUID: owner.ID,
+			Label:           label,
+		}
+		if err := st.upsertSecretApplicationOwner(ctx, tx, dbSecretOwner); err != nil {
+			return errors.Annotatef(err, "updating application secret record for secret %q", uri)
+		}
+	case coresecrets.UnitOwner:
+		dbSecretOwner := secretUnitOwner{
+			SecretID: uri.ID,
+			// Query selects the unit uuid as owner id.
+			UnitUUID: owner.ID,
+			Label:    label,
+		}
+		if err := st.upsertSecretUnitOwner(ctx, tx, dbSecretOwner); err != nil {
+			return errors.Annotatef(err, "updating unit secret record for secret %q", uri)
+		}
+	}
+	return nil
+}
+
+func (st *State) markObsoleteRevisions(ctx context.Context, tx *sqlair.TX, uri *coresecrets.URI, keepRevision int) error {
+	query := `
+UPDATE secret_revision
+SET    obsolete = True
+WHERE  revision != $M.revision
+AND    secret_id = $M.secret_id
+AND    revision NOT IN (
+           SELECT current_revision FROM secret_unit_consumer
```

Comment:
I don't think this is correct.
The secret_unit_consumer and secret_remote_unit_consumer don't have current_revision field.
I am implementing the markObsoleteRevisions in my watcher PR.
So no need to fix we can just delete it.